### PR TITLE
Logger Interface

### PR DESCRIPTION
**Pull request by pjdufour-truss**
## Description

This PR migrates our use of zap.Logger `type` to an internal logger `interface` for each package.  This new approach provides something similar to a `contract` that explicitly defines each packages logging 
requirements.  This approach takes advantage of Go's feature that doesn't require explicit `implements` or `extends` of interfaces.  This feature allows us to use interfaces internal to a package that simplify our inter-dependencies, though they do require defining our logger interface more than once for each package.  This adds some marginal cost, but is much more scalable for a growing codebase. 

Additionally, since we are now using an interface, we can make project-specific enhancements to our loggers that may not be suitable for merging upstream with [zap](https://godoc.org/go.uber.org/zap).

Functionally, this PR allows us to pass the `git_branch` and `git_commit` fields through the request cycle and will provide hooks to add additional logging fields over time.  For example, we'd like to log `task id`, `session id`, and a unique `request id` with every relevant log message so we can more easily search our application logs.

## Reviewer Notes

- Should we make the `logger` interface for each package public?  While not "required", private interfaces aren't shown by `godoc`.
- How do we document this pattern?

## Setup

None

## Code Review Verification Steps

* [ ] Code follows the guidelines for [Logging](./docs/backend.md#logging)
* [ ] The requirements listed in
 [Querying the Database Safely](./docs/backend.md#querying-the-database-safely)
 have been satisfied.
* Any new migrations/schema changes:
  * [ ] Follow our guidelines for zero-downtime deploys (see [Zero-Downtime Deploys](./docs/database.md#zero-downtime-migrations))
  * [ ] Have been communicated to #dp3-engineering
  * [ ] Secure migrations have been tested using `bin/run-prod-migrations`
* [ ] There are no aXe warnings for UI.
* [ ] This works in IE.
* Any new client dependencies (Google Analytics, hosted libraries, CDNs, etc) have been:
  * [ ] Communicated to @willowbl00
  * [ ] Added to the list of [network dependencies](https://github.com/transcom/mymove#client-network-dependencies)
* [ ] Tested in the Experimental environment (for changes to containers, app startup, or connection to data stores)
* [ ] Request review from a member of a different team.
* [ ] Have the Pivotal acceptance criteria been met for this change?

## References

* [Pivotal story](https://www.pivotaltracker.com/story/show/163604570) for this change

## Screenshots

N.A.

### CODE REVIEW

File: pkg/uploader/uploader_test.go [hunk 1]
```diff
@@ -21,7 +20,7 @@ import (
 
 type UploaderSuite struct {
 	testingsuite.PopTestSuite
-	logger       *zap.Logger
+	logger       logint.Logger
```
Comment (chrisgilmerproj) [inline pkg/uploader/uploader_test.go]: What is `logint`?

File: cmd/webserver/main.go [hunk 2]
```diff
@@ -681,13 +681,25 @@ func main() {
...
 
+	var logger logger
 	zapLogger, err := logging.Config(env, v.GetBool("debug-logging"))
 	if err != nil {
 		log.Fatalf("Failed to initialize Zap logging due to %v", err)
 	}
 	zap.ReplaceGlobals(zapLogger)
 
-	logger := &webserverLogger{zapLogger}
+	if len(gitBranch) > 0 || len(gitCommit) > 0 {
```
Comment (mikena-truss) [inline cmd/webserver/main.go]: Can this entire block can be:
```
        fields := make([]zap.Field, 0)
		if len(gitBranch) > 0 {
			fields = append(fields, zap.String("git_branch", gitBranch))
		}
		if len(gitCommit) > 0 {
			fields = append(fields, zap.String("git_commit", gitCommit))
		}
		logger = zapLogger.With(fields...)
```

You'll automatically make the slice, but `With` handles that just fine and it saves some readability and `len` calls

File: pkg/handlers/logger.go [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+type logger interface {
+	Debug(msg string, fields ...zap.Field)
+	Info(msg string, fields ...zap.Field)
+	Error(msg string, fields ...zap.Field)
+	Warn(msg string, fields ...zap.Field)
+	Fatal(msg string, fields ...zap.Field)
+}
+
+// WithOptions wraps zap.WithOptions
+func withOptions(logger logger, opts ...zap.Option) logger {
```
Comment (mikena-truss) [inline pkg/handlers/logger.go]: Why wrap this instead of adding it to the interface?

File: cmd/webserver/main.go (hunk 2 above)
Comment (pjdufour-truss) [reply cmd/webserver/main.go]: Yeah, looks like `With` just returns the object pointer without copying.  Good idea.

https://github.com/uber-go/zap/blob/master/logger.go#L159

File: pkg/handlers/logger.go (hunk 3 above)
Comment (pjdufour-truss) [reply pkg/handlers/logger.go]: 👍

File: cmd/webserver/logger.go [hunk 4]
```diff
@@ -0,0 +1,13 @@
+package main
+
+import (
+	"go.uber.org/zap"
+)
+
+type logger interface {
+	Debug(msg string, fields ...zap.Field)
```
Comment (rdhariwal) [inline cmd/webserver/logger.go]: Would we consider simplifying the signature for debug/error etc functions one more step? We can use a string map like this `Debug(msg string, arguments map[string] string)` and that should remove zap library dependency. I realize that might have a chain reaction effect of changes in all handlers where logger is already using zap but that should remove one more dependency for us.

Comment (rdhariwal) [reply cmd/webserver/logger.go]: on the other hand I do see how zap can handle not just string but other types.. so take this comment with a grain of salt

**Comment by rdhariwal** [comment]
Looks great to me overall, great job! I would like to add my 2 cents to what @mikena-truss has already mentioned about having a single interface across packages for the `logger` so that developers don't have to translate which "version" of the logger they are working with and DI will work very cleanly if there is just once interface.

Having said that you can change me mind :)

File: cmd/webserver/logger.go (hunk 4 above)
Comment (pjdufour-truss) [reply cmd/webserver/logger.go]: You're in my mind, personally, I think that's a good objective, but want to pave the road there first.

**Pull request merged**